vrx_gz/src/GymkhanaScoringPlugin.cc

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <gz/msgs/param.pb.h>
+#include <gz/msgs/vector3d.pb.h>
+#include <limits>
+#include <gz/sim/Entity.hh>
+#include <gz/sim/Model.hh>
+#include <gz/sim/Util.hh>
+#include <gz/sim/World.hh>
+#include <gz/plugin/Register.hh>
+#include <gz/transport/Node.hh>
+#include <sdf/sdf.hh>
+
+#include "GymkhanaScoringPlugin.hh"
+
+using namespace gz;
+using namespace vrx;
+
+/// \brief Private GymkhanaScoringPlugin data class.
+class GymkhanaScoringPlugin::Implementation
+{
+  /// \brief Callback for black box station-keeping portion's scoring plugin
+  public: void BlackboxCallback(const msgs::Param &_msg);
+
+  /// \brief Set the pinger location.
+  public: void SetPingerPosition();
+
+  /// \brief Transport node
+  public: transport::Node node;
+
+  /// \brief ROS publisher to set the pinger position.
+  public: transport::Node::Publisher setPingerPub;
+
+  /// \brief Whether buoy channel portion has been completed
+  public: bool channelCrossed = false;
+
+  /// \brief Cumulative error from black box station keeping portion
+  public: double blackboxScore = 0.0;
+
+  /// \brief Penalty added per collision.
+  public: double obstaclePenalty = 0.1;
+
+  // TODO: compare with acousticpinger
+  /// \brief Position of the pinger.
+  public: math::Vector3d pingerPosition = {0, 0, 0};
+
+  /// \brief Last time we published a pinger position.
+  public: std::chrono::duration<double> lastSetPingerPositionTime = 
+    std::chrono::duration<double>::zero();
+
+  /// \brief Pointer to the SDF plugin element.
+  public: sdf::ElementPtr sdf;
+
+  /// \brief Display or suppress state changes
+  public: bool silent = false;
+
+};
+
+/////////////////////////////////////////////////
+void GymkhanaScoringPlugin::ChannelCallback(
+  const msgs::Param &_msg)
+{
+  if (_msg.IsInitialized()) // TODO: Check
```

Comment:
Out of curiosity, what's the reason for checking this?

Reply:
I don't know... maybe it isn't necessary? (That's what my TODO is reminding me to ask about.) I was trying to port the code from our previous implementation in `gymkhana_scoring_plugin.cc`, which uses:
```
if (msg)
```
I assumed the intent was to check whether a null message had been passed into the callback, but the syntax produces an error in gazebosim, which will not automatically cast msg as a boolean, so I added `IsInitialized()` thinking this was the intent. As to why the callback would be receiving an empty message... I don't know. Seems like a workaround for a bug, maybe.  Should I take it out?

Reply:
I don't think this is necessary here. In the past maybe `_msg` was a pointer and we wanted to make sure it wasn't `null` but it's safe here.